### ISSUE TITLE
Power Laws in Deep Learning

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#394]
Power Laws in Deep Learning<br>
<br>
https://ift.tt/2x0fdrp<br>
<br>
Charles H Martin, PhD<br>
<br>

<div>
<div>

		<p><a href="https://stats385.github.io/" target="_blank" rel="noopener">Why Does Deep Learning Work ? </a>   If we could get a better handle on this, we could solve some very hard problems in Deep Learning. A new class at Stanford is trying to address this. Here, I review some very new results <a href="https://www.youtube.com/watch?v=_Ni5UDrVwYU&amp;" target="_blank" rel="noopener">from my own research (w/UC Berkeley)</a> on the topic which has direct practical application.  Here, I will show that</p>
<p><strong><span>In pretrained, production quality DNNs,  the weight matrices for the Fully Connected (FC ) layers display Fat Tailed Power Law behavior.</span></strong></p>
<h3><span>Setup: the Energy Landscape function</span></h3>
<p>Deep Neural Networks (DNNs) minimize the Energy function defined by their architecture.  We define the layer weight matrices <img src="https://s0.wp.com/latex.php?latex=%5Cmathbf%7BW_%7BL%7D%7D+&amp;bg=ffffff&amp;fg=303030&amp;s=0" alt="\mathbf{W_{L}} " title="\mathbf{W_{L}} " class="latex">, biases, <img src="https://s0.wp.com/latex.php?latex=%5Cmathbf%7Bb_%7BL%7D%7D+&amp;bg=ffffff&amp;fg=303030&amp;s=0" alt="\mathbf{b_{L}} " title="\mathbf{b_{L}} " class="latex">, and activations functions <img src="https://s0.wp.com/latex.php?latex=h_%7BL%7D+&amp;bg=ffffff&amp;fg=303030&amp;s=0" alt="h_{L} " title="h_{L} " class="latex">, giving</p>
<p><img src="https://s0.wp.com/latex.php?latex=E_%7BDNN%7D%3Dh_%7BL%7D%28%5Cmathbf%7BW%7D_%7BL%7D%5Ctimes+h_%7BL-1%7D%28%5Cmathbf%7BW%7D_%7BL-1%7D%5Ctimes+h_%7BL-2%7D%28%5Ccdots%29%2B%5Cmathbf%7Bb%7D_%7BL-1%7D%29%2B%5Cmathbf%7Bb%7D_%7BL%7D%29+&amp;bg=ffffff&amp;fg=303030&amp;s=0" alt="E_{DNN}=h_{L}(\mathbf{W}_{L}\times h_{L-1}(\mathbf{W}_{L-1}\times h_{L-2}(\cdots)+\mathbf{b}_{L-1})+\mathbf{b}_{L}) " title="E_{DNN}=h_{L}(\mathbf{W}_{L}\times h_{L-1}(\mathbf{W}_{L-1}\times h_{L-2}(\cdots)+\mathbf{b}_{L-1})+\mathbf{b}_{L}) " class="latex"></p>
<p>We train the DNN on a labeled data set (d,y), giving the optimization problem</p>
<p><img src="https://s0.wp.com/latex.php?latex=%5Cunderset%7BW_%7Bl%7D%2Cb_%7Bl%7D%7D%7B%5Cmbox%7Bmin%7D%7D%5C%3B%5Cmathcal%7BL%7D%5Cleft%28%5Csum_%7Bi%7DE_%7BDNN%7D%28d_%7Bi%7D%29-y_%7Bi%7D%29%5Cright%29+&amp;bg=ffffff&amp;fg=303030&amp;s=0" alt="\underset{W_{l},b_{l}}{\mbox{min}}\;\mathcal{L}\left(\sum_{i}E_{DNN}(d_{i})-y_{i})\right) " title="\underset{W_{l},b_{l}}{\mbox{min}}\;\mathcal{L}\left(\sum_{i}E_{DNN}(d_{i})-y_{i})\right) " class="latex"></p>
<p>We call this the Energy Landscape because the DNN optimization problem is only parameterized by the weights and biases.  Of course, in any real DNN problem, we do have other adjustable parameters,  such as the amount of Dropout, the learning rate, the batch size, etc.   But these regularization effects simply change the global <span><a href="https://calculatedcontent.com/2015/03/25/why-does-deep-learning-work/" target="_blank" rel="noopener"><img src="https://s0.wp.com/latex.php?latex=E_%7BDNN%7D+&amp;bg=ffffff&amp;fg=303030&amp;s=0" alt="E_{DNN} " title="E_{DNN} " class="latex"> </a></span></p>
<p><span>The Energy Landscape function</span><span> changes on each epoch–and do we care about how. In fact, <a href="https://calculatedcontent.com/2015/03/25/why-does-deep-learning-work/" target="_blank" rel="noopener">I have argued that <img src="https://s0.wp.com/latex.php?latex=E_%7BDNN%7D+&amp;bg=ffffff&amp;fg=303030&amp;s=0" alt="E_{DNN} " title="E_{DNN} " class="latex"> must form an Energy Funnel:</a> </span></p>
<figure data-shortcode="caption" id="attachment_12718" class="wp-caption aligncenter"><img data-attachment-id="12718" data-permalink="https://calculatedcontent.com/2018/09/09/power-laws-in-deep-learning/screen-shot-2018-09-09-at-8-58-26-pm-2/" data-orig-file="https://charlesmartin14.files.wordpress.com/2018/09/screen-shot-2018-09-09-at-8-58-26-pm1.png" data-orig-size="1020,656" data-comments-opened="1" data-image-meta='{"aperture":"0","credit":"","camera":"","caption":"","created_timestamp":"0","copyright":"","focal_length":"0","iso":"0","shutter_speed":"0","title":"","orientation":"0"}' data-image-title="Screen Shot 2018-09-09 at 8.58.26 PM" data-medium-file="https://charlesmartin14.files.wordpress.com/2018/09/screen-shot-2018-09-09-at-8-58-26-pm1.png?w=300" data-large-file="https://charlesmartin14.files.wordpress.com/2018/09/screen-shot-2018-09-09-at-8-58-26-pm1.png?w=840" src="https://charlesmartin14.files.wordpress.com/2018/09/screen-shot-2018-09-09-at-8-58-26-pm1.png?w=292&amp;h=188" alt="Screen Shot 2018-09-09 at 8.58.26 PM.png"><figcaption class="wp-caption-text">Energy Funnel</figcaption></figure>
<p>But here, for now, we only look at the final result.  Once a DNN is trained, what is left are the weights (and biases).   We can reuse the weights (and biases) of a pre-trained DNN to build new DNNs with transfer learning. And if we train a bunch of DNNs, we want to know which one is better ?</p>
<p><span>But, practically, we would really like to identify a very good DNN <em>without peaking at the test data</em>, since every time we peak, and retrain, we risk overtraining our DNN.</span></p>
<p>I now show we can at least start do this by looking the weights matrices themselves.  So let us look at the weights of some pre-trained DNNs.</p>
<h4><span>Pretrained Linear Models in Pytorch</span></h4>
<p>Pytorch comes with several pretrained models, such as AlexNet. To start, we just examine the weight matrices of the Linear / Fully Connected (FC) layers.</p>
<pre>
pretrained_model = models.alexnet(pretrained=True)
for module in pretrained_model.modules():
     if isinstance(module, nn.Linear):
     ...
</pre>
<p>The Linear layers have the simplest weight matrices <img src="https://s0.wp.com/latex.php?latex=%5Cmathbf%7BW%7D_%7Bl%7D+&amp;bg=ffffff&amp;fg=303030&amp;s=0" alt="\mathbf{W}_{l} " title="\mathbf{W}_{l} " class="latex"> ; they are 2-dimensional tensors, or just rectangular matrices.</p>
<p>Let <img src="https://s0.wp.com/latex.php?latex=%5Cmathbf%7BW%7D_%7Bl%7D+&amp;bg=ffffff&amp;fg=303030&amp;s=0" alt="\mathbf{W}_{l} " title="\mathbf{W}_{l} " class="latex"> be an <img src="https://s0.wp.com/latex.php?latex=M%5Ctimes+N+&amp;bg=ffffff&amp;fg=303030&amp;s=0" alt="M\times N " title="M\times N " class="latex"> matrix, where <img src="https://s0.wp.com/latex.php?latex=M%5Cle+N+&amp;bg=ffffff&amp;fg=303030&amp;s=0" alt="M\le N " title="M\le N " class="latex">.  We can get the matrix from the pretraing model using:</p>
<pre>
W = np.array(module.weight.data.clone().cpu())
M, N = np.min(W.shape), np.max(W.shape)
</pre>
<h3>Empirical Spectral Density</h3>
<p>How is information concentrated in <img src="https://s0.wp.com/latex.php?latex=%5Cmathbf%7BW%7D_%7Bl%7D+&amp;bg=ffffff&amp;fg=303030&amp;s=0" alt="\mathbf{W}_{l} " title="\mathbf{W}_{l} " class="latex">.  For any rectangular matrix, we can form the</p>
<h4>Singular Value Decomposition (SVD)</h4>
<p><img src="https://s0.wp.com/latex.php?latex=%5Cmathbf%7BW%7D_%7Bl%7D%3D%5Cmathbf%7BU%7D%5Cmathbf%7B%5CSigma%7D%5Cmathbf%7BV%7D%5ET+&amp;bg=ffffff&amp;fg=303030&amp;s=0" alt="\mathbf{W}_{l}=\mathbf{U}\mathbf{\Sigma}\mathbf{V}^T " title="\mathbf{W}_{l}=\mathbf{U}\mathbf{\Sigma}\mathbf{V}^T " class="latex"></p>
<p><img src="https://s0.wp.com/latex.php?latex=%5Cnu_%7Bi%7D%3D%5CSigma_%7Bii%7D%2C%5C%3B%5C%3Bi%3D1%5Ccdots+M+&amp;bg=ffffff&amp;fg=303030&amp;s=0" alt="\nu_{i}=\Sigma_{ii},\;\;i=1\cdots M " title="\nu_{i}=\Sigma_{ii},\;\;i=1\cdots M " class="latex"></p>
<p>which is readily computed in scikit learn. We will use the faster TruncatedSVD method, and compute singular values <img src="https://s0.wp.com/latex.php?latex=%5Cnu_%7Bi%7D+&amp;bg=ffffff&amp;fg=303030&amp;s=0" alt="\nu_{i} " title="\nu_{i} " class="latex">:</p>
<pre>
from sklearn.decomposition import TruncatedSVD
svd = TruncatedSVD(n_components=M-1, n_iter=7, random_state=10)
svd.fit(W)
svals = svd.singular_values_
</pre>
<p>(Technically, we do miss the smallest singular value doing this, but that’s ok.  It won’t matter here, and we can always use the pure svd method to be a exact)</p>
<h4>Eigenvalues of the Correlation Matrix</h4>
<p>We can, alternatively form the eigenvalues of the correlation matrix</p>
<p><img src="https://s0.wp.com/latex.php?latex=%5Cmathbf%7BX%7D+%3D+%5Cdfrac%7B1%7D%7BN%7D%5Cmathbf%7BW%7D_%7Bl%7D%5E%7BT%7D%5Cmathbf%7BW%7D_%7Bl%7D++&amp;bg=ffffff&amp;fg=303030&amp;s=0" alt="\mathbf{X} = \dfrac{1}{N}\mathbf{W}_{l}^{T}\mathbf{W}_{l}  " title="\mathbf{X} = \dfrac{1}{N}\mathbf{W}_{l}^{T}\mathbf{W}_{l}  " class="latex"></p>
<p><img src="https://s0.wp.com/latex.php?latex=%5Cmathbf%7BX%7D%5Cmathbf%7Bv%7D_%7Bi%7D%3D%5Clambda_%7Bi%7D%5Cmathbf%7Bv%7D_%7Bi%7D+&amp;bg=ffffff&amp;fg=303030&amp;s=0" alt="\mathbf{X}\mathbf{v}_{i}=\lambda_{i}\mathbf{v}_{i} " title="\mathbf{X}\mathbf{v}_{i}=\lambda_{i}\mathbf{v}_{i} " class="latex"></p>
<p>The eigenvalues are just the square of the singular values.</p>
<p><img src="https://s0.wp.com/latex.php?latex=%5Clambda_%7Bi%7D%3D%5Cdfrac%7B1%7D%7BN%7D%5Cnu%5E%7B2%7D_%7Bi%7D+&amp;bg=ffffff&amp;fg=303030&amp;s=0" alt="\lambda_{i}=\dfrac{1}{N}\nu^{2}_{i} " title="\lambda_{i}=\dfrac{1}{N}\nu^{2}_{i} " class="latex"></p>
<p>Notice here we normalize them by N.</p>
<pre>
evals = (1/N)*svals*svals
</pre>
<p>We now form the Empirical Spectral Density (ESD), which is, formally</p>
<p><img src="https://s0.wp.com/latex.php?latex=%5Crho_%7Bemp%7D%28%5Clambda%29%3D%5Cdfrac%7B1%7D%7BN%7D%5Csum_%7Bi%3D1%7D%5E%7BM%7D+%5Cdelta%5Cleft%28+%5Clambda+-+%5Clambda_i%5Cright%29+&amp;bg=ffffff&amp;fg=303030&amp;s=0" alt="\rho_{emp}(\lambda)=\dfrac{1}{N}\sum_{i=1}^{M} \delta\left( \lambda - \lambda_i\right) " title="\rho_{emp}(\lambda)=\dfrac{1}{N}\sum_{i=1}^{M} \delta\left( \lambda - \lambda_i\right) " class="latex"></p>
<p>This notation just means compute a histogram of the eigenvalues</p>
<pre>
import matplotlib.pyplot as plt
plt.hist(evals, bins=100, density=True)
</pre>
<p>We could also compute the spectral density using a Kernel Density Estimator (KDE); we save this for a future post.</p>
<p>We now look at the ESD of</p>
<h4>Pretrained AlexNet</h4>
<p>Here, we examine just FC3, the last Linear layer, connecting the model to the labels.  The other linear layers, FC1 and FC2, look similar.  Below is a histogram for ESD.  Notice it is very sharply peaked and has long tail, extending out past 40.</p>
<p><a href="https://calculatedcontent.com/2018/09/09/power-laws-in-deep-learning/alexnet-fc3-full-2/" rel="attachment wp-att-12569"><img data-attachment-id="12569" data-permalink="https://calculatedcontent.com/2018/09/09/power-laws-in-deep-learning/alexnet-fc3-full-2/" data-orig-file="https://charlesmartin14.files.wordpress.com/2018/08/alexnet-fc3-full1.png" data-orig-size="360,360" data-comments-opened="1" data-image-meta='{"aperture":"0","credit":"","camera":"","caption":"","created_timestamp":"0","copyright":"","focal_length":"0","iso":"0","shutter_speed":"0","title":"","orientation":"0"}' data-image-title="alexnet-fc3-full" data-medium-file="https://charlesmartin14.files.wordpress.com/2018/08/alexnet-fc3-full1.png?w=300&amp;h=300" data-large-file="https://charlesmartin14.files.wordpress.com/2018/08/alexnet-fc3-full1.png?w=360" class="wp-image-12569 size-medium aligncenter" src="https://charlesmartin14.files.wordpress.com/2018/08/alexnet-fc3-full1.png?w=300&amp;h=300"></a></p>
<p>We can get a better view of the heavy tailed behavior by zooming in.</p>
<p><a href="https://calculatedcontent.com/2018/09/09/power-laws-in-deep-learning/alexnet-fc3-2/" rel="attachment wp-att-12570"><img data-attachment-id="12570" data-permalink="https://calculatedcontent.com/2018/09/09/power-laws-in-deep-learning/alexnet-fc3-2/" data-orig-file="https://charlesmartin14.files.wordpress.com/2018/08/alexnet-fc31.png" data-orig-size="360,360" data-comments-opened="1" data-image-meta='{"aperture":"0","credit":"","camera":"","caption":"","created_timestamp":"0","copyright":"","focal_length":"0","iso":"0","shutter_speed":"0","title":"","orientation":"0"}' data-image-title="alexnet-fc3" data-medium-file="https://charlesmartin14.files.wordpress.com/2018/08/alexnet-fc31.png?w=300&amp;h=300" data-large-file="https://charlesmartin14.files.wordpress.com/2018/08/alexnet-fc31.png?w=360" class="wp-image-12570 size-medium aligncenter" src="https://charlesmartin14.files.wordpress.com/2018/08/alexnet-fc31.png?w=300&amp;h=300"></a></p>
<p>The red curve is a fit of the ESD to the Marchenko Pastur (MP) Random Matrix Theory (RMT) result — it is not a very good fit.  This means ESD does not resemble Gaussian Random matrix.  Instead, it is looks heavy tailed.  Which leads to the question…</p>
<h3>Do we have a Power Law ?</h3>
<p>(Yes do, as we shall see…)</p>
<p>Physicists love to claim they have discovered data that follows a power law:</p>
<p><img src="https://s0.wp.com/latex.php?latex=P%28%5Clambda%29%5Csim%5Cdfrac%7B1%7D%7B%5Clambda%5E%7B%5Calpha%7D%7D%2C%5C%3B%5C%3B%5Calpha%3E0+&amp;bg=ffffff&amp;fg=303030&amp;s=0" alt="P(\lambda)\sim\dfrac{1}{\lambda^{\alpha}},\;\;\alpha&gt;0 " title="P(\lambda)\sim\dfrac{1}{\lambda^{\alpha}},\;\;\alpha&gt;0 " class="latex"></p>
<p>But this is harder to do than it seems.  <a href="http://www.stat.cmu.edu/~cshalizi/2010-10-18-Meetup.pdf">And statisticians love to point this out.  </a>Don’t be fooled–we physicists knew this;  <a href="https://www.springer.com/us/book/9783540308829">Sornette’s book</a> has a whole chapter on it.  Still, we have to use best practices.</p>
<h4>Log Log Histograms</h4>
<p>The first thing to do: plot the data on a log-log histogram, and check that this plot is linear–at least in some region.  Let’s look at our ESD for AlexNet FC3:</p>
<figure data-shortcode="caption" id="attachment_12589" class="wp-caption aligncenter"><a href="https://calculatedcontent.com/2018/09/09/power-laws-in-deep-learning/alexnet-log-log-fc3-2/" rel="attachment wp-att-12589"><img data-attachment-id="12589" data-permalink="https://calculatedcontent.com/2018/09/09/power-laws-in-deep-learning/alexnet-log-log-fc3-2/" data-orig-file="https://charlesmartin14.files.wordpress.com/2018/09/alexnet-log-log-fc31.png" data-orig-size="360,360" data-comments-opened="1" data-image-meta='{"aperture":"0","credit":"","camera":"","caption":"","created_timestamp":"0","copyright":"","focal_length":"0","iso":"0","shutter_speed":"0","title":"","orientation":"0"}' data-image-title="alexnet-log-log-fc3" data-medium-file="https://charlesmartin14.files.wordpress.com/2018/09/alexnet-log-log-fc31.png?w=300&amp;h=300" data-large-file="https://charlesmartin14.files.wordpress.com/2018/09/alexnet-log-log-fc31.png?w=360" class="wp-image-12589 size-medium" src="https://charlesmartin14.files.wordpress.com/2018/09/alexnet-log-log-fc31.png?w=300&amp;h=300"></a><figcaption class="wp-caption-text">AlexNet FC3: Log log histogram of the Empirical Spectral Density (ESD). Notice the central region is mostly linear, suggesting a power law.</figcaption></figure>
<p>Yes, it is linear–in the central region, for eigenvalue frequencies between roughly ~1 and ~100–and that is most of the distribution.</p>
<p>Why is not linear everywhere?  Because it is finite size–there are min and max cutoffs.  In the infinite limit, a powerlaw <img src="https://s0.wp.com/latex.php?latex=p%28x%29%C2%A0+&amp;bg=ffffff&amp;fg=303030&amp;s=0" alt="p(x)  " title="p(x)  " class="latex"> diverges at <img src="https://s0.wp.com/latex.php?latex=x%3D0++&amp;bg=ffffff&amp;fg=303030&amp;s=0" alt="x=0  " title="x=0  " class="latex"> , and the tail extends indefinitely as <img src="https://s0.wp.com/latex.php?latex=x%5Crightarrow%5Cinfty++&amp;bg=ffffff&amp;fg=303030&amp;s=0" alt="x\rightarrow\infty  " title="x\rightarrow\infty  " class="latex">. In any finite size data set, there will be an <img src="https://s0.wp.com/latex.php?latex=x_%7Bmin%7D++&amp;bg=ffffff&amp;fg=303030&amp;s=0" alt="x_{min}  " title="x_{min}  " class="latex"> and <img src="https://s0.wp.com/latex.php?latex=x_%7Bmax%7D++&amp;bg=ffffff&amp;fg=303030&amp;s=0" alt="x_{max}  " title="x_{max}  " class="latex">.</p>
<h4>Power Law Fits</h4>
<p>Second, fit the data to a power law, with <img src="https://s0.wp.com/latex.php?latex=x_%7Bmin%7D++&amp;bg=ffffff&amp;fg=303030&amp;s=0" alt="x_{min}  " title="x_{min}  " class="latex"> and <img src="https://s0.wp.com/latex.php?latex=x_%7Bmax%7D++&amp;bg=ffffff&amp;fg=303030&amp;s=0" alt="x_{max}  " title="x_{max}  " class="latex"> in mind.  The most widely available and accepted method the Maximum Likelihood Estimator (MLE), develop by <a href="https://arxiv.org/abs/0706.1062">Clauset et. al</a>., and available in the <a href="https://github.com/jeffalstott/powerlaw">python powerlaw package</a>.</p>
<pre>
import powerlaw
fit = powerlaw.Fit(evals, xmax=np.max(evals))
alpha, D = fit.alpha, fit.D
</pre>
<p> </p>
<p>The D value is a quality metric, the KS distance.   There are other options as well.  The smaller D, the better.  The table below shows typical values of good fits.</p>
<h4>Linear Log Log Plots</h4>
<p>The powerlaw package also makes some great plots.  Below is a log log plot generated for our fit of FC3, for the central region of the ESD.  The filled lines represent our fits, and the dotted  lines are actual power lawPDF (<span>blue</span>) and CCDF (<span>red</span>).   The filled lines look like straight lines and overlap the dotted lines–so this fit looks pretty good.</p>
<figure data-shortcode="caption" id="attachment_12592" class="wp-caption aligncenter"><a href="https://calculatedcontent.com/2018/09/09/power-laws-in-deep-learning/screen-shot-2018-09-01-at-9-50-07-am/" rel="attachment wp-att-12592"><img data-attachment-id="12592" data-permalink="https://calculatedcontent.com/2018/09/09/power-laws-in-deep-learning/screen-shot-2018-09-01-at-9-50-07-am/" data-orig-file="https://charlesmartin14.files.wordpress.com/2018/09/screen-shot-2018-09-01-at-9-50-07-am.png" data-orig-size="728,640" data-comments-opened="1" data-image-meta='{"aperture":"0","credit":"","camera":"","caption":"","created_timestamp":"0","copyright":"","focal_length":"0","iso":"0","shutter_speed":"0","title":"","orientation":"0"}' data-image-title="Screen Shot 2018-09-01 at 9.50.07 AM" data-medium-file="https://charlesmartin14.files.wordpress.com/2018/09/screen-shot-2018-09-01-at-9-50-07-am.png?w=300&amp;h=264" data-large-file="https://charlesmartin14.files.wordpress.com/2018/09/screen-shot-2018-09-01-at-9-50-07-am.png?w=728" class="wp-image-12592 size-medium" src="https://charlesmartin14.files.wordpress.com/2018/09/screen-shot-2018-09-01-at-9-50-07-am.png?w=300&amp;h=264"></a><figcaption class="wp-caption-text">AlexNet FC3: Log Log plot of the central region of the Empirical Spectral Density and best Powerlaw fits: pdf (blue) and ccdf (red)</figcaption></figure>
<p>Is this enough ?  Not yet…</p>
<h4>Quality of the Estimator</h4>
<p>We still need to know, do we have enough data to get a good estimate for <img src="https://s0.wp.com/latex.php?latex=%5Calpha+&amp;bg=ffffff&amp;fg=303030&amp;s=0" alt="\alpha " title="\alpha " class="latex">, what are our error bars, and what kind of systematic errors might we get?</p>
<ol>
<li>This so-called statistically valid MLE estimator actually only works properly for <img src="https://s0.wp.com/latex.php?latex=%5Calpha%5Cin%5B1.5%2C3.5%5D+&amp;bg=ffffff&amp;fg=303030&amp;s=0" alt="\alpha\in[1.5,3.5] " title="\alpha\in[1.5,3.5] " class="latex">.</li>
<li>And we need a lot of data points…more than we have singular values <img src="https://s0.wp.com/latex.php?latex=%5Cnu_%7Bi%7D+&amp;bg=ffffff&amp;fg=303030&amp;s=0" alt="\nu_{i} " title="\nu_{i} " class="latex">
</li>
</ol>
<p>We can calibrate the estimator by generating some modest size (N=1000) random power law datasets using the numpy Pareto function, where</p>
<p><img src="https://s0.wp.com/latex.php?latex=%5Calpha%3D%5Cmu%2B1+&amp;bg=ffffff&amp;fg=303030&amp;s=0" alt="\alpha=\mu+1 " title="\alpha=\mu+1 " class="latex"></p>
<p>and then fitting these with the PowerLaw package.  We get the following curve<a href="https://calculatedcontent.com/2018/09/09/power-laws-in-deep-learning/alexnet-log-log-fc3-overlay/" rel="attachment wp-att-12493"> </a></p>
<figure data-shortcode="caption" id="attachment_12539" class="wp-caption aligncenter"><a href="https://calculatedcontent.com/2018/09/09/power-laws-in-deep-learning/powerlaw-calibration/" rel="attachment wp-att-12539"><img data-attachment-id="12539" data-permalink="https://calculatedcontent.com/2018/09/09/power-laws-in-deep-learning/powerlaw-calibration/" data-orig-file="https://charlesmartin14.files.wordpress.com/2018/08/powerlaw-calibration.png" data-orig-size="360,360" data-comments-opened="1" data-image-meta='{"aperture":"0","credit":"","camera":"","caption":"","created_timestamp":"0","copyright":"","focal_length":"0","iso":"0","shutter_speed":"0","title":"","orientation":"0"}' data-image-title="Powerlaw calibration curve" data-medium-file="https://charlesmartin14.files.wordpress.com/2018/08/powerlaw-calibration.png?w=300&amp;h=300" data-large-file="https://charlesmartin14.files.wordpress.com/2018/08/powerlaw-calibration.png?w=360" class="wp-image-12539 size-medium" src="https://charlesmartin14.files.wordpress.com/2018/08/powerlaw-calibration.png?w=300&amp;h=300"></a><figcaption class="wp-caption-text">Powerlaw fits of exponent compared to data generated from a Pareto distribution, N=1000. Notice that the estimator works best for exponents between 1.5 and 3.5</figcaption></figure>
<p>The green line is a perfect estimate.  The Powerlaw package overestimates small <img src="https://s0.wp.com/latex.php?latex=%5Calpha%3C1.5+&amp;bg=ffffff&amp;fg=303030&amp;s=0" alt="\alpha&lt;1.5 " title="\alpha&lt;1.5 " class="latex"> and underestimates large <img src="https://s0.wp.com/latex.php?latex=%5Calpha%3E3.5+&amp;bg=ffffff&amp;fg=303030&amp;s=0" alt="\alpha&gt;3.5 " title="\alpha&gt;3.5 " class="latex">. Fortunately, most of our fits lie in the good range.</p>
<h4>Is a Power law the best model ?</h4>
<p>A good fit is not enough. We also should ensure that no other obvious model is a better fit. The power law package lets us test out fit against other common (long tailed) choices, namely</p>
<ul>
<li>exponential (EXP)</li>
<li>stretched_exponential (S_EXP)</li>
<li>log normal  (LOG_N)</li>
<li>truncated power law (TPL)</li>
</ul>
<p>For example, to check if our data is better fit by a log normal distribution, we run</p>
<pre>

R, p = fit.distribution_compare('powerlaw', 'lognormal', normalized_ratio=True)

</pre>
<p>and R and the  the p-value.  If if R&lt;0 and p &lt;= 0.05, then we can conclude that a power law is a better model.</p>
<p>Note that sometimes, for <img src="https://s0.wp.com/latex.php?latex=%5Calpha%5Cleq2+&amp;bg=ffffff&amp;fg=303030&amp;s=0" alt="\alpha\leq2 " title="\alpha\leq2 " class="latex">, the best model may be a truncated power law (TPL).  This happens because our data sets are pretty small, and the tails of our ESDs fall off pretty fast.   A TPL is just a power law (PL) with an exponentially decaying tail, and it may seem to be a better fit for small data, fixed size sets.  Also, the TPL also has 2 parameters, whereas the PL has only 1, so it is not unexpected that the TPL would fit the data better.</p>
<h4>More Pretrained Models:  Linear Layers</h4>
<p>Below we fit the linear layers of <a href="https://pytorch.org/docs/stable/torchvision/models.html" target="_blank" rel="noopener">the many pretrained models in pytorch</a>.  All fit a power law (PL) or truncated power law (TPL)</p>
<figure data-shortcode="caption" id="attachment_12587" class="wp-caption aligncenter"><a href="https://calculatedcontent.com/2018/09/09/power-laws-in-deep-learning/power-law-exponents/" rel="attachment wp-att-12587"><img data-attachment-id="12587" data-permalink="https://calculatedcontent.com/2018/09/09/power-laws-in-deep-learning/power-law-exponents/" data-orig-file="https://charlesmartin14.files.wordpress.com/2018/09/power-law-exponents.png" data-orig-size="360,360" data-comments-opened="1" data-image-meta='{"aperture":"0","credit":"","camera":"","caption":"","created_timestamp":"0","copyright":"","focal_length":"0","iso":"0","shutter_speed":"0","title":"","orientation":"0"}' data-image-title="power-law-exponents" data-medium-file="https://charlesmartin14.files.wordpress.com/2018/09/power-law-exponents.png?w=300&amp;h=300" data-large-file="https://charlesmartin14.files.wordpress.com/2018/09/power-law-exponents.png?w=360" class="wp-image-12587 size-medium" src="https://charlesmartin14.files.wordpress.com/2018/09/power-law-exponents.png?w=300&amp;h=300"></a><figcaption class="wp-caption-text">Powerlaw exponents for ESD of Fully Connected / Linear layers of numerous pretrained ImageNet models currently available in pytorch.</figcaption></figure>
<p>The table below lists all the fits, with the Best Fit and the KS statistic D.  <a href="https://calculatedcontent.com/2018/09/09/power-laws-in-deep-learning/screen-shot-2018-08-31-at-8-29-25-pm/" rel="attachment wp-att-12486"><img data-attachment-id="12486" data-permalink="https://calculatedcontent.com/2018/09/09/power-laws-in-deep-learning/screen-shot-2018-08-31-at-8-29-25-pm/" data-orig-file="https://charlesmartin14.files.wordpress.com/2018/08/screen-shot-2018-08-31-at-8-29-25-pm.png?w=840&amp;h=962" data-orig-size="1402,1606" data-comments-opened="1" data-image-meta='{"aperture":"0","credit":"","camera":"","caption":"","created_timestamp":"0","copyright":"","focal_length":"0","iso":"0","shutter_speed":"0","title":"","orientation":"0"}' data-image-title="Power Law Fits for Pretrained Models in Pytorch" data-medium-file="https://charlesmartin14.files.wordpress.com/2018/08/screen-shot-2018-08-31-at-8-29-25-pm.png?w=840&amp;h=962?w=262" data-large-file="https://charlesmartin14.files.wordpress.com/2018/08/screen-shot-2018-08-31-at-8-29-25-pm.png?w=840&amp;h=962?w=840" src="https://charlesmartin14.files.wordpress.com/2018/08/screen-shot-2018-08-31-at-8-29-25-pm.png?w=840&amp;h=962"></a></p>
<p><a href="https://calculatedcontent.com/2018/09/09/power-laws-in-deep-learning/screen-shot-2018-08-31-at-8-29-25-pm/" rel="attachment wp-att-12486"><span>Notice that all the power law exponents lie in the range 2-4, except for a couple outliers.</span></a></p>
<p><a href="https://calculatedcontent.com/2018/09/09/power-laws-in-deep-learning/screen-shot-2018-08-31-at-8-29-25-pm/" rel="attachment wp-att-12486"><span>   <img src="https://s0.wp.com/latex.php?latex=%5Calpha%5Cin%5B2%2C4%5D+&amp;bg=ffffff&amp;fg=303030&amp;s=0" alt="\alpha\in[2,4] " title="\alpha\in[2,4] " class="latex"></span></a></p>
<p>This is pretty fortunate since out PL estimator only works well around this range.  This is also pretty remarkable because it suggests there is…</p>
<h4>Universality in Deep Learning:</h4>
<p>There is a deep connection between this range of exponents and some (relatively) recent results in Random Matrix Theory (RMT).   Indeed, this seems to suggest that Deep Learning systems display the kind of Universality seen in Self Organized systems, like <a href="https://www.nature.com/articles/srep35831">real spiking neurons</a>.  We will examine this in a future post.</p>
<h4>Practical Applications:</h4>
<p>Training DNNs is hard.  There are many tricks one has to use.  And you have to monitor the training process carefully.</p>
<p>Here we suggest that the ESDs of the Linear / FC layers of a well trained DNN will display power law behavior. And the exponents will be between 2 and 4.</p>
<h4>Counterexample: Inception V3</h4>
<p>A counter example is Inception V3.  Both FC layers 226 and 302 have unusually large <img src="https://s0.wp.com/latex.php?latex=%5Calpha&amp;bg=ffffff&amp;fg=303030&amp;s=0" alt="\alpha" title="\alpha" class="latex"> exponents. Looking closer at Layer 222,  the ESDs is not a power law at all, but rather bi-model heavy tailed distribution.   <a href="https://calculatedcontent.com/2018/09/09/power-laws-in-deep-learning/inception-l1/" rel="attachment wp-att-12666"><img data-attachment-id="12666" data-permalink="https://calculatedcontent.com/2018/09/09/power-laws-in-deep-learning/inception-l1/" data-orig-file="https://charlesmartin14.files.wordpress.com/2018/09/inception-l1.png?w=840" data-orig-size="360,360" data-comments-opened="1" data-image-meta='{"aperture":"0","credit":"","camera":"","caption":"","created_timestamp":"0","copyright":"","focal_length":"0","iso":"0","shutter_speed":"0","title":"","orientation":"0"}' data-image-title="inception-l1" data-medium-file="https://charlesmartin14.files.wordpress.com/2018/09/inception-l1.png?w=840?w=300" data-large-file="https://charlesmartin14.files.wordpress.com/2018/09/inception-l1.png?w=840?w=360" src="https://charlesmartin14.files.wordpress.com/2018/09/inception-l1.png?w=840"></a></p>
<p>We conjecture that as good as Inception V3 is, perhaps it could be further optimized.  It would be interesting to see if anyone can show this.</p>
<h4>Appendix:</h4>
<h4>Derivation of the MLE PowerLaw estimator</h4>
<p>The commonly accepted method uses Maximum Likelihood Estimator (MLE).  This</p>
<p><img src="https://s0.wp.com/latex.php?latex=p%28x%3B%5Calpha%29%5Csim+x%5E%7B-%5Calpha%7D+&amp;bg=ffffff&amp;fg=303030&amp;s=0" alt="p(x;\alpha)\sim x^{-\alpha} " title="p(x;\alpha)\sim x^{-\alpha} " class="latex"></p>
<p><img src="https://s0.wp.com/latex.php?latex=p%28x%3B%5Calpha%29%3D%5Cdfrac%7B%5Calpha-1%7D%7Bx_%7Bmin%7D%7D%5Cleft%28%5Cdfrac%7Bx%7D%7Bx_%7Bmin%7D%7D%5Cright%29%5E%7B-%5Calpha%7D+&amp;bg=ffffff&amp;fg=303030&amp;s=0" alt="p(x;\alpha)=\dfrac{\alpha-1}{x_{min}}\left(\dfrac{x}{x_{min}}\right)^{-\alpha} " title="p(x;\alpha)=\dfrac{\alpha-1}{x_{min}}\left(\dfrac{x}{x_{min}}\right)^{-\alpha} " class="latex"></p>
<p><img src="https://s0.wp.com/latex.php?latex=P%28X%29%3D%5Cint_%7Bx%7D%5E%7B%5Cinfty%7Dp%28x%27%29dx%27%3D%5Cleft%28%5Cdfrac%7Bx%7D%7Bx_%7Bmin%7D%7D%5Cright%29%5E%7B-%5Calpha%2B1%7D+&amp;bg=ffffff&amp;fg=303030&amp;s=0" alt="P(X)=\int_{x}^{\infty}p(x')dx'=\left(\dfrac{x}{x_{min}}\right)^{-\alpha+1} " title="P(X)=\int_{x}^{\infty}p(x')dx'=\left(\dfrac{x}{x_{min}}\right)^{-\alpha+1} " class="latex"></p>
<p>The conditional probability is defined by evaluating <img src="https://s0.wp.com/latex.php?latex=p%28x%3B%5Calpha%29+&amp;bg=ffffff&amp;fg=303030&amp;s=0" alt="p(x;\alpha) " title="p(x;\alpha) " class="latex"> on a data set (of size n)</p>
<p><img src="https://s0.wp.com/latex.php?latex=p%28x%7C%5Calpha%29%3D%5Cdisplaystyle%5Cprod_%7Bi%3D1%7D%5E%7Bn%7D%5Cdfrac%7B%5Calpha-1%7D%7Bx_%7Bmin%7D%7D%5Cleft%28%5Cdfrac%7Bx_%7Bi%7D%7D%7Bx_%7Bmin%7D%7D%5Cright%29%5E%7B-%5Calpha%7D+&amp;bg=ffffff&amp;fg=303030&amp;s=0" alt="p(x|\alpha)=\displaystyle\prod_{i=1}^{n}\dfrac{\alpha-1}{x_{min}}\left(\dfrac{x_{i}}{x_{min}}\right)^{-\alpha} " title="p(x|\alpha)=\displaystyle\prod_{i=1}^{n}\dfrac{\alpha-1}{x_{min}}\left(\dfrac{x_{i}}{x_{min}}\right)^{-\alpha} " class="latex"></p>
<p>The log likelihood is</p>
<p><img src="https://s0.wp.com/latex.php?latex=%5Cmathcal%7BL%7D%3Dln%5C%3Bp%28x%7C%5Calpha%29%3D%5Cln%5C%3B%5Cdisplaystyle%5Cprod_%7Bi%3D1%7D%5E%7Bn%7D%5Cdfrac%7B%5Calpha-1%7D%7Bx_%7Bmin%7D%7D%5Cleft%28%5Cdfrac%7Bx_%7Bi%7D%7D%7Bx_%7Bmin%7D%7D%5Cright%29%5E%7B-%5Calpha%7D+&amp;bg=ffffff&amp;fg=303030&amp;s=0" alt="\mathcal{L}=ln\;p(x|\alpha)=\ln\;\displaystyle\prod_{i=1}^{n}\dfrac{\alpha-1}{x_{min}}\left(\dfrac{x_{i}}{x_{min}}\right)^{-\alpha} " title="\mathcal{L}=ln\;p(x|\alpha)=\ln\;\displaystyle\prod_{i=1}^{n}\dfrac{\alpha-1}{x_{min}}\left(\dfrac{x_{i}}{x_{min}}\right)^{-\alpha} " class="latex"></p>
<p>Which is easily reduced to</p>
<p><img src="https://s0.wp.com/latex.php?latex=%3D%5Cdisplaystyle%5Csum_%7Bi%3D1%7D%5E%7Bn%7D%5CBig%5Bln%28%5Calpha-1%29-ln%28x_%7Bmin%7D%29-%5Calpha%5C%3Bln%5Cleft%28%5Cdfrac%7Bx_%7Bi%7D%7D%7Bx_%7Bmin%7D%7D%5Cright%29%5CBig%5D+&amp;bg=ffffff&amp;fg=303030&amp;s=0" alt="=\displaystyle\sum_{i=1}^{n}\Big[ln(\alpha-1)-ln(x_{min})-\alpha\;ln\left(\dfrac{x_{i}}{x_{min}}\right)\Big] " title="=\displaystyle\sum_{i=1}^{n}\Big[ln(\alpha-1)-ln(x_{min})-\alpha\;ln\left(\dfrac{x_{i}}{x_{min}}\right)\Big] " class="latex"></p>
<p><img src="https://s0.wp.com/latex.php?latex=%3Dn%5C%3Bln%28%5Calpha-1%29-n%5C%3B+ln%28x_%7Bmin%7D%29-%5Calpha%5C%3B+ln%5Cdisplaystyle%5Csum_%7Bi%3D1%7D%5E%7Bn%7D%5Cleft%28%5Cdfrac%7Bx_%7Bi%7D%7D%7Bx_%7Bmin%7D%7D%5Cright%29+&amp;bg=ffffff&amp;fg=303030&amp;s=0" alt="=n\;ln(\alpha-1)-n\; ln(x_{min})-\alpha\; ln\displaystyle\sum_{i=1}^{n}\left(\dfrac{x_{i}}{x_{min}}\right) " title="=n\;ln(\alpha-1)-n\; ln(x_{min})-\alpha\; ln\displaystyle\sum_{i=1}^{n}\left(\dfrac{x_{i}}{x_{min}}\right) " class="latex"></p>
<p>The maximum likelihood occurs when</p>
<p><img src="https://s0.wp.com/latex.php?latex=%5Cdfrac%7B%5Cpartial%5Cmathcal%7BL%7D%7D%7B%5Cpartial%5Calpha%7D%3D0+&amp;bg=ffffff&amp;fg=303030&amp;s=0" alt="\dfrac{\partial\mathcal{L}}{\partial\alpha}=0 " title="\dfrac{\partial\mathcal{L}}{\partial\alpha}=0 " class="latex"></p>
<p>The MLE estimate, for a given <img src="https://s0.wp.com/latex.php?latex=x_%7Bmin%7D+&amp;bg=ffffff&amp;fg=303030&amp;s=0" alt="x_{min} " title="x_{min} " class="latex">, is:</p>
<p><img src="https://s0.wp.com/latex.php?latex=%5Chat%7B%5Calpha%7D%3D1%2Bn%5CBig%5B%5Cdisplaystyle%5Csum_%7Bi%3D1%7D%5E%7Bn%7Dln%5Cdfrac%7Bx_%7Bi%7D%7D%7Bx_%7Bmin%7D%7D%5CBig%5D%5E%7B-1%7D++&amp;bg=ffffff&amp;fg=303030&amp;s=0" alt="\hat{\alpha}=1+n\Big[\displaystyle\sum_{i=1}^{n}ln\dfrac{x_{i}}{x_{min}}\Big]^{-1}  " title="\hat{\alpha}=1+n\Big[\displaystyle\sum_{i=1}^{n}ln\dfrac{x_{i}}{x_{min}}\Big]^{-1}  " class="latex"></p>
<p>We can either input <img src="https://s0.wp.com/latex.php?latex=x_%7Bmin%7D+&amp;bg=ffffff&amp;fg=303030&amp;s=0" alt="x_{min} " title="x_{min} " class="latex">, or search for the best possible estimate (<a href="https://arxiv.org/abs/0706.1062" target="_blank" rel="noopener">as explained in the Clauset <em>et. al.</em> paper</a>).</p>
<p>Notice, however, that this estimator does not explicitly take <img src="https://s0.wp.com/latex.php?latex=x_%7Bmax%7D+&amp;bg=ffffff&amp;fg=303030&amp;s=0" alt="x_{max} " title="x_{max} " class="latex"> into account.  And for this reason it seems to be very limited in its application. A better method, such as <a href="https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0170920" target="_blank" rel="noopener">developed recently by Thurner </a>(but not yet coded in python), may prove more robust for a larger range of exponents.</p>
<h4>Extending this to Convolutional (Conv2D) Layers:</h4>
<p>We can naively extend these results to Conv2D layers by reshaping the Conv2D 4-D Tensors back into rectangular 2D matrices.  Doing this, and repeating the fits, we see very different results:</p>
<p><a href="https://calculatedcontent.com/2018/09/09/power-laws-in-deep-learning/screen-shot-2018-09-02-at-9-08-03-pm/" rel="attachment wp-att-12667"><img data-attachment-id="12667" data-permalink="https://calculatedcontent.com/2018/09/09/power-laws-in-deep-learning/screen-shot-2018-09-02-at-9-08-03-pm/" data-orig-file="https://charlesmartin14.files.wordpress.com/2018/09/screen-shot-2018-09-02-at-9-08-03-pm.png" data-orig-size="746,674" data-comments-opened="1" data-image-meta='{"aperture":"0","credit":"","camera":"","caption":"","created_timestamp":"0","copyright":"","focal_length":"0","iso":"0","shutter_speed":"0","title":"","orientation":"0"}' data-image-title="Screen Shot 2018-09-02 at 9.08.03 PM" data-medium-file="https://charlesmartin14.files.wordpress.com/2018/09/screen-shot-2018-09-02-at-9-08-03-pm.png?w=300&amp;h=271" data-large-file="https://charlesmartin14.files.wordpress.com/2018/09/screen-shot-2018-09-02-at-9-08-03-pm.png?w=746" src="https://charlesmartin14.files.wordpress.com/2018/09/screen-shot-2018-09-02-at-9-08-03-pm.png?w=300&amp;h=271"></a></p>
<p>Many Conv2D layers show very high exponents.  What could this mean?</p>
<ol>
<li>The PL estimator is terrible for large exponents. We need to implement a more reliable PL estimator before we can start looking at Conv2D  layers.</li>
<li>We reshaped the Conv2D layers in a naive way.  Perhaps there is some more clever way to reshare the Conv2D Tensors into 2D arrays that better reflects the underlying math</li>
<li>Or, perhaps it is simply that the Linear layers, being fully connected, are also fully correlated, whereas the Conv2D layers do less work.</li>
</ol>
<p>It will be very exciting to find out.</p>
<h4>Jupyter Notebooks</h4>
<p><a href="https://github.com/CalculatedContent/ImplicitSelfRegularization" target="_blank" rel="noopener">Code for this study, <em>and future work,</em> is available in this repo</a>.</p>
<h4>Video Presentations</h4>
<p><a href="https://www.youtube.com/c/calculationconsulting" target="_blank" rel="noopener">Research talks are available on my youtube channel.  </a><span>Please Subscribe</span></p>

		
	</div>

	</div><br>
<br>
<br>
<br>
AI|ML|DM|IR|PR<br>
<br>
via Machine Learning https://ift.tt/2knH5Pn<br>
<br>
September 10, 2018 at 01:05PM
